flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,194 @@
+		this.backingShuffleEnvironment = Preconditions.checkNotNull(backingShuffleEnvironment);
+	}
+
+	@Override
+	public boolean hasPartitionsOccupyingLocalResources(JobID jobId) {
+		return inProgressPartitionTable.hasTrackedPartitions(jobId) || finishedPartitionTable.hasTrackedPartitions(jobId);
+	}
+
+	@Override
+	public void markJobActive(JobID jobId) {
+		activeJobs.add(jobId);
+	}
+
+	@Override
+	public void releaseFinishedPartitions(JobID jobId, Collection<ResultPartitionID> resultPartitionIds) {
+		finishedPartitionTable.stopTrackingPartitions(jobId, resultPartitionIds);
+		backingShuffleEnvironment.releasePartitions(resultPartitionIds);
+	}
+
+	@Override
+	public void releaseAllFinishedPartitionsForJobAndMarkJobInactive(JobID jobId) {
+		activeJobs.remove(jobId);
+		Collection<ResultPartitionID> finishedPartitionsForJob = finishedPartitionTable.stopTrackingPartitions(jobId);
+		backingShuffleEnvironment.releasePartitions(finishedPartitionsForJob);
+	}
+
+	/**
+	 * This method wraps partition writers for externally managed partitions and introduces callbacks into the lifecycle
+	 * methods of the {@link ResultPartitionWriter}.
+	 */
+	@Override
+	public Collection<NotifyingResultPartitionWriter> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup) {
+
+		final Collection<? extends ResultPartitionWriter> resultPartitionWriters = backingShuffleEnvironment
+			.createResultPartitionWriters(taskName, executionAttemptID, resultPartitionDeploymentDescriptors, outputGroup, buffersGroup);
+
+		// determine which partitions are externally managed, and whose writers need to be wrapped
+		final Set<ResultPartitionID> externallyManagedPartitions = resultPartitionDeploymentDescriptors.stream()
+			.filter(ResultPartitionDeploymentDescriptor::isManagedExternally)
+			// TODO: The descriptor should be extended to also contain the ResultPartitionID
+			.map(descriptor -> new ResultPartitionID(descriptor.getPartitionId(), executionAttemptID))
+			.collect(Collectors.toSet());
+
+		// wrap partition writers to introduce callbacks for lifecycle events
+		final Collection<NotifyingResultPartitionWriter> wrappedResultPartitionWriters = new ArrayList<>(resultPartitionWriters.size());
+		for (ResultPartitionWriter resultPartitionWriter : resultPartitionWriters) {
+			final NotifyingResultPartitionWriter wrapper;
+			if (externallyManagedPartitions.contains(resultPartitionWriter.getPartitionId())) {
+				wrapper = new NotifyingResultPartitionWriter(
+					resultPartitionWriter,
+					partitionId -> notifyPartitionSetup(jobId, partitionId),
+					partitionId -> notifyPartitionFailed(jobId, partitionId),
+					partitionId -> notifyPartitionFinished(jobId, partitionId));
+			} else {
+				wrapper = new NotifyingResultPartitionWriter(
```

Comment:
If not externally managed, another option is not wrapping the partition, and only use the original partition, as long as we define the interface as `Collection<ResultPartitionWriter> createResultPartitionWriters`. WDYT?

Reply:
huh, I thought I tried that..., will adjust it.